umbraco/Umbraco-CMS
# Fixes packager to ensure the GUIDs are used for all entities where possible

When package data is installed we want to make sure that it installs using the GUIDs assigned/exported so that new entities are not created when the package is reinstalled. This is the case for many entities but for others it isn't.

* DataTypes = OK
* Languages = By ISO Code = OK (no GUID in the DB)
* Dictionary = NOT OK, only looked up by the string key but needs to use the GUID instead
* Macros = NOT OK, only looked up by string alias but needs to use the GUID
* Template = OK, looked up by alias because there is no GUID in the DB
* Doc types = NOT OK, only looked up by alias but needs to use the GUID
* Media types = NOT OK, only looked up by alias but needs to use the GUID
* Stylesheets = OK, uses the name, this is just a file anyways
* Documents = OK, uses the 'key' (GUID) but it seems to be optional so lets make that not optional (see CreateContentFromXml)
* Media = OK, uses the 'key' (GUID) but it seems to be optional so lets make that not optional (see CreateContentFromXml)
* Macro properties = NOT OK
* Property types = NOT OK

## Testing

This is a little annoying but here's what to do:

* Create a new doc type with one property
* Create a 2nd nested doc type with one property (to ensure that the nesting still works with GUIDs)
* Create a media type with one property
* Create a couple nested dictionary items
* Create a macro with a property
* Then create a package with all of these specific items selected
* Then export the package.xml file
* Then go delete all of these created items in Umbraco

In your webhost testing project (i.e. `Umbraco.Web.UI.NetCore`), Create a folder/namespace like `MyPackage`. Then add an automatic package migration there like:

```cs
public class TestAutomaticPackageMigrationPlan : AutomaticPackageMigrationPlan
{
    public TestAutomaticPackageMigrationPlan() : base("MyPackage")
    {
    }
}
```

The move your downloaded package.xml file into this folder and mark as an embedded resource.

Then start your project. This will run the automatic package migration and install all of this data. You'll then need to go into the DB to lookup the items created in the various tables like umbracoNode (for doc/media types), cmsPropertyType (for property types), cmsMacro (for macros), cmsMacroProperty (for macro properties), cmsLanguageText (for dictionary items). And then verify all of the GUIDs are the same as all of the GUIDs in your package.xml file.

Then stop the site and make a small change to the file, like removing an empty `<Validation />` node. This will change the hash of this file which is used to create the automatic package migration plan Id. So then start the site again and it will run another package migration for the same file. Then you can verify that it didn't make any duplicates and all the GUIDs are still the same.
